### PR TITLE
Storages: support building vector index for ColumnFileTiny (Part 1)

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #9032

Problem Summary:

### What is changed and how it works?

This PR is part of https://github.com/tidbcloud/tiflash-cse/pull/293, which mainly adds index info in ColumnFileTiny meta.

```commit-message

```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by breezewish — APPROVED** [review]

File: dbms/src/Storages/DeltaMerge/tests/gtest_column_file_clone.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+
+namespace DB::DM::tests
+{
+
+class DeltaValueSpaceTest : public SegmentTestBasic
+{
+};
+
+TEST_F(DeltaValueSpaceTest, CloneColumnFileTinyWithVectorIndex)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/tests/gtest_column_file_clone.cpp:29]:
```suggestion
class ColumnFileCloneTest : public SegmentTestBasic
{
};

TEST_F(ColumnFileCloneTest, CloneColumnFileTinyWithVectorIndex)
```

File: dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.h [hunk 2]
```diff
@@ -34,9 +35,18 @@ using ColumnFileTinyPtr = std::shared_ptr<ColumnFileTiny>;
...
 class ColumnFileTiny : public ColumnFilePersisted
 {
+public:
     friend class ColumnFileTinyReader;
     friend struct Remote::Serializer;
 
+    struct IndexInfo
+    {
+        PageIdU64 index_page_id{};
+        std::optional<dtpb::VectorIndexFileProps> vector_index = std::nullopt;
```
Comment (breezewish) [inline dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.h:45]: In what cases will we have an empty index page?

File: dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.cpp [hunk 3]
```diff
@@ -236,36 +256,55 @@ ColumnFilePersistedPtr ColumnFileTiny::restoreFromCheckpoint(
...
     auto column_file_schema = std::make_shared<ColumnFileSchema>(*schema);
-    return std::make_shared<ColumnFileTiny>(column_file_schema, rows, bytes, new_cf_id, context);
+    if (!index_infos)
+        return std::make_shared<ColumnFileTiny>(column_file_schema, rows, bytes, new_cf_id, context);
+
+    // Write index data page to local ps
+    auto new_index_infos = std::make_shared<IndexInfos>();
+    for (const auto & index : *index_infos)
+    {
+        auto new_index_page_id = put_remote_page(index.index_page_id);
```
Comment (CalvinNeo) [inline dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.cpp:301]: How to ensure that the page must exist?

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_Statistics.cpp [hunk 4]
```diff
@@ -231,42 +231,66 @@ LocalIndexesStats DeltaMergeStore::getLocalIndexStats()
...
+            // Delta
             {
-                const auto [state, bytes] = dmfile->getLocalIndexState(index_info.column_id, index_info.index_id);
-                UNUSED(bytes);
-                switch (state)
+                const auto & delta = segment->getDelta();
+                const auto & mem_table = delta->getMemTableSet();
+                index_stats.rows_delta_not_indexed += mem_table->getRows();
+                const auto & persisted = delta->getPersistedFileSet();
+                for (const auto & file : persisted->getFiles())
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_Statistics.cpp:240]: This for-loop maybe not thread-safe?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore_Statistics.cpp:243]: Fixed.

File: dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.h [hunk 5]
```diff
@@ -87,17 +100,29 @@ class ColumnFileTiny : public ColumnFilePersisted
...
+        const IndexInfosPtr & index_infos_ = nullptr,
         const CachePtr & cache_ = nullptr);
 
     Type getType() const override { return Type::TINY_FILE; }
 
     size_t getRows() const override { return rows; }
     size_t getBytes() const override { return bytes; }
 
+    IndexInfosPtr getIndexInfos() const { return index_infos; }
+    bool hasIndex(Int64 index_id) const
```
Comment (CalvinNeo) [inline dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.h:112]: `hasVectorIndex`?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.h:112]: `hasIndex` is better since we will support other types of index later.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**